runtime/gc_structs/ClassArrayClassSlotIterator.cpp

```diff
@@ -36,32 +36,40 @@
  * @return the next slot containing an object reference
  * @return NULL if there are no more such slots
  */
-J9Class **
+J9Class *
 GC_ClassArrayClassSlotIterator::nextSlot() 
 {
-	J9Class **slotPtr;
+	J9Class *classPtr;
 
 	switch(_state) {
 	case classArrayClassSlotIterator_state_arrayClass:
-		slotPtr = (J9Class **)&_iterateClazz->arrayClass;
+		classPtr = (J9Class *)_iterateClazz->arrayClass;
 		if(!_isArrayClass) {
 			_state = classArrayClassSlotIterator_state_done;
 		} else {
 			_state += 1;
 		}
-		break;
+		if (NULL != classPtr) {
+			return classPtr;
+		}
```

Comment:
I think this might be not correct for case it is not an array class. `_state` is set to `classArrayClassSlotIterator_state_done` and if `classPtr` is NULL code switches to next state `classArrayClassSlotIterator_state_componentType` and this is not applicable for non-array class.
I am wondering what if instead of removing breaks you put entire `switch` to the `while` loop and let state machine work normally.

Reply:
Ah this helps. I was trying to pinpoint what was wrong in the array class iterator. I'll try that.

Reply:
how about a check at the end of loop (do-while style)? it would be less checks